feat: constraint weight override

triceo opened the pull request:
Planning solution gets an optional field of type `Map` (or more likely a custom map), with keys being the `constraintName` from `asConstraint(constraintName)`, and the value being the constraint weight. 
Using the map is optional. Constraint weights not present in the map use their default constraint weights as defined by `penalize(...)`. Constraints present in the map but not present in the `ConstraintProvider` trigger a fail-fast.

`asConstraintDescribed(constraintName, constraintDescription)` is introduced to keep constraint documentation where it belongs, with the constraints.

`ConstraintWeight`, `ConstraintConfiguration` and `penalizeConfigurable(...)` are deprecated. They continue working, but the new system doesn't use them.

triceo commented:
@lee-carlon This PR will continue evolving, but the docs are likely already stable. As always, feedback is welcome.

lee-carlon commented on docs/src/modules/ROOT/pages/constraints-and-score/constraint-configuration.adoc:
```diff
@@ -1,21 +1,139 @@
+...
+----
+
+The `Vehicle capacity` constraint in this planning solution will get a weight of `2hard`,
```
```suggestion
The `Vehicle capacity` constraint in this planning solution has a weight of `2hard`,
```

lee-carlon commented on docs/src/modules/ROOT/pages/constraints-and-score/constraint-configuration.adoc:
```diff
@@ -1,21 +1,139 @@
+
+The `Vehicle capacity` constraint in this planning solution will get a weight of `2hard`,
+as opposed to its original `1hard`.
+The `Service finished after max end time` constraint will have a weight of `0hard`,
```
```suggestion
The `Service finished after max end time` constraint has a weight of `0hard`,
```

lee-carlon commented on docs/src/modules/ROOT/pages/constraints-and-score/constraint-configuration.adoc:
```diff
@@ -1,21 +1,139 @@
+The `Service finished after max end time` constraint will have a weight of `0hard`,
+and therefore will be disabled entirely.
+
+This way, you can solve the same problem
+while applying different constraint weights to each instance.
```
```suggestion
In this way, you can solve the same problem
by applying different constraint weights to each instance.
```

lee-carlon reviewed (approved):
@triceo looks good. Just a few minor suggestions.

zepfred commented on core/src/main/java/ai/timefold/solver/core/api/score/stream/tri/TriConstraintStream.java:
```diff
@@ -1892,10 +1915,10 @@ default Constraint penalizeBigDecimal(String constraintPackage, String constrain
      * @return never null
+     * @deprecated Prefer {@link #penalize(Score, ToIntTriFunction)} and {@link ConstraintWeightOverrides}.
      */
     @Deprecated(forRemoval = true)
```
Should we include version `1.13.0` and update other stream classes?

triceo replied:
Not sure here. I left it here like this, because it was already deprecated before. Saying it's "since 1.13.0" wouldn't be true, because it was deprecated years ago already.

zepfred commented on core/src/main/java/ai/timefold/solver/core/impl/domain/solution/ConstraintWeightSupplier.java:
```diff
@@ -0,0 +1,53 @@
+import ai.timefold.solver.core.impl.domain.common.accessor.MemberAccessorFactory;
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+
+public sealed interface ConstraintWeightSupplier<Solution_, Score_ extends Score<Score_>>
```
Nice refactoring!

zepfred commented on migration/src/main/java/ai/timefold/solver/migration/v8/AsConstraintRecipe.java:
```diff
@@ -237,9 +238,8 @@ public Expression visitExpression(Expression expression, ExecutionContext execut
-                            LOGGER.warn("Cannot refactor to asConstraint() method" +
-                                    " for deprecated method called in expression (" + e + ").");
-                            return e;
+                            LOGGER.warn("Cannot refactor to asConstraint() method for deprecated method (" + method + ").");
```
```suggestion
                            LOGGER.warn("Cannot refactor to asConstraint() method for deprecated method ({}).", method);
```

zepfred reviewed (approved):
Great job. The new configuration mechanism looks much better!

zepfred commented:
It looks like the Python module is broken.

triceo commented:
Python needs a follow-up PR. I didn't provide one yet, because we're in the middle of moving repositories.

triceo commented:
@Christopher-Chianelli I've made several Python changes:

- Removed several deprecated methods.
- Added equivalent to `asConstraintDescribed(...)`.

I need your review on those.

I also attempted (but not finished) to redefine `ConstraintWeightOverrides` in Python. (Using the native Java one didn't work, as it kept complaining of incompatible generics on Java `HashMap`.) Please finish this work, so that I could learn from it.

triceo commented on python/python-core/src/main/python/domain/_types.py:
```diff
@@ -0,0 +1,18 @@
+
+Score_ = TypeVar('Score_', bound='Score')
+
+class ConstraintWeightOverrides(Generic[Score_]):
```
Needs finishing.

triceo commented on python/python-core/tests/test_solver_configuration.py:
```diff
@@ -5,8 +5,7 @@
 from typing import Annotated, List
-
-
+"""
```
Needs enabling once overrides work.

triceo commented:
Thanks, @Christopher-Chianelli!
